lib/python/bins.cpp

```diff
@@ -138,35 +138,36 @@ void init_buckets(py::module &m) {
   m.def("is_bins",
         [](const Dataset &dataset) { return dataset::is_bins(dataset); });
 
-  m.def("bins_begin_end", [](const Variable &var) -> py::object {
-    if (var.dtype() == dtype<bucket<Variable>>)
-      return bin_begin_end<Variable>(var);
-    if (var.dtype() == dtype<bucket<DataArray>>)
-      return bin_begin_end<DataArray>(var);
-    if (var.dtype() == dtype<bucket<Dataset>>)
-      return bin_begin_end<Dataset>(var);
-    return py::none();
-  });
-
-  m.def("bins_dim", [](const Variable &var) -> py::object {
+  m.def(
+      "bins_begin_end",
+      [](const Variable &var) -> std::optional<std::pair<Variable, Variable>> {
+        if (var.dtype() == dtype<bucket<Variable>>)
+          return bin_begin_end<Variable>(var);
+        if (var.dtype() == dtype<bucket<DataArray>>)
+          return bin_begin_end<DataArray>(var);
+        if (var.dtype() == dtype<bucket<Dataset>>)
+          return bin_begin_end<Dataset>(var);
+        return std::nullopt;
```

Comment:
Maybe we should just throw here instead of returning an optional? I can't see the use of this now.

Reply:
Or should we expose `constituents` directly? It looks like this function and the ones below are only used to implement `constituents` in Python.

Reply:
I think that may indeed be possible now. The code here is likely a leftover from quite recent times when we could not get the "begin" and "end" indices without allocating two large new variables. I think I had therefore split this into multiple functions, such that this cost could be avoided if only the dim or buffer were required.